README.md
<img src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/addon/icon128.png" align="right">

Salesforce inspector
===========================
Chrome and Firefox extension to add a metadata layout on top of the standard Salesforce UI to improve the productivity and joy of Salesforce configuration, development, and integration work.

[![Test Status](https://travis-ci.org/tprouvot/Chrome-Salesforce-inspector.svg?branch=master)](https://travis-ci.org/tprouvot/Chrome-Salesforce-inspector)

New Features / Fix
-----
[List of changes](CHANGES.md)

Features
-----
* Quickly view field information directly from a record detail page, edit page or Visualforce page.
* Quickly view and edit all data for a record, even data that is not on the page layout.
* Perform quick one-off data exports and imports directly from within Salesforce. Data can be easily copied to and from Excel. No need to log in again when you are already logged in with your browser.
* View current limits consumption
* Simple access to user record details incl. simpler access to "login as". E.g. when hopping between test users.
* And more...

<img alt="Inspector menu" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/1.png" height="100">
<img alt="Show field metadata" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/2.png" height="100">
<img alt="Show all data for record" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/3.png" height="100">
<img alt="Data exporter" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/4.png" height="100">
<img alt="Data importer" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/5.png" height="100">
<img alt="Monitor limits" src="https://raw.githubusercontent.com/sorenkrabbe/Chrome-Salesforce-inspector/master/docs/screenshots/6.png" height="100">

Security and Privacy
-----
The Salesforce Inspector browser extension/plugin communicates directly between the user's web browser and the Salesforce servers. No data is sent to other parties and no data is persisted outside of Salesforce servers after the user leaves the Salesforce Inspector pages.
The Inspector communicates via the official Salesforce webservice APIs on behalf of the currently logged in user. This means the Inspector will be capable of accessing nothing but the data and features the user has been granted access to in Salesforce.

All Salesforce API calls from the Inspector re-uses the access token/session used by the browser to access Salesforce. To acquire this access token the Salesforce Inspector requires permission to read browser cookie information for Salesforce domains.

To validate the accuracy of this description, inspect the source code, monitor the network traffic in your browser or take my word.

Use Salesforce Inspector with a Connected App
-----
If you enabled "API client whitelisting", Sf Inspector may not work anymore.
To secure the extension usage, you can use a auth flow to get an access token linked to a connected app.

1. Create a connected app.
2. Set permissions and callback url. (chrome-extension://chromeExtensionId/data-export.html?host=mysandboxHost&)
3. Get client Id and save it in the export page

<img alt="Connected App" src="./docs/screenshots/connectedApp.png" height="300">
<img alt="Client Id" src="./docs/screenshots/clientId.png" height="100">
<img alt="Generate Token" src="./docs/screenshots/generateAccessToken.png" height="300">

Installation
-----
1. Download or clone the repo locally.
2. Open `chrome://extensions/`.
3. Enable `Developer mode`.
4. Click `Load unpacked extension...`.
5. Select the `addon` subdirectory of this repository.

Troubleshooting
-----
* If Salesforce Inspector is not available after installation, the most likely issue is that your browser is not up to date. See [instructions for Google Chrome](https://productforums.google.com/forum/#!topic/chrome/YK1-o4KoSjc).
* When you enable the My Domain feature in Salesforce, Salesforce Inspector may not work until you have restarted your browser (or until you have deleted the "sid" cookie for the old Salesforce domain by other means).

Development
-----

1. Install Node.js with npm
2. `npm install`

### Chrome
1. `npm run chrome-dev-build`
2. Open `chrome://extensions/`.
3. Enable `Developer mode`.
4. Click `Load unpacked extension...`.
5. Select the `addon` subdirectory of this repository.

### Firefox

1. `npm run firefox-dev-build`
2. In Firefox, open `about:debugging`.
3. Click `Load Temporary Add-on…`.
4. Select the file `addon/manifest.json`.

Unit tests
-----
1. Set up an org (e.g. a Developer Edition) and apply the following customizations:
   1.a. Everything described in metadata in `test/org/`. Push to org with `sfdx force:mdapi:deploy --deploydir test/org -w 1000 -u [your-test-org-alias]`
   1.b. Ensure _Allow users to relate a contact to multiple accounts_ is enabled (Setup→Account Settings)
   1.c. Ensure the org has no _namespace prefix_ (Setup→Package Manager)
2. Navigate to one of the extension pages and replace the file name with `test-framework.html`, for example `chrome-extension://example/test-framework.html?host=example.my.salesforce.com`.
3. Wait until "Salesforce Inspector unit test finished successfully" is shown.
4. If the test fails, open your browser's developer tools console to see error messages.

### Linting

1. `npm run eslint`

Release
-------
Version number must be manually incremented in [addon/manifest-template.json](addon/manifest-template.json) file

### Chrome

If the version number is greater than the version currently in Chrome Web Store, the revision will be packaged and uploaded to the store ready for manual release to the masses.

### Firefox

1. `npm run firefox-release-build`
2. Upload the file from `target/firefox/firefox-release-build.zip` to addons.mozilla.org

Design Principles
-----
(we don't live up to all of them. pull requests welcome)
* Stay completely inactive until the user explicitly interacts with it. The tool has the potential to break Salesforce functionality when used, since we rely on monkey patching and internal APIs. We must ensure that you cannot break Salesforce just by having the tool installed or enabled. For example, we won't fix the setup search placeholder bug.
* For manual ad-hoc tasks only. The tool is designed to help administrators and developers interact with Salesforce in the browser. It is after all a browser add-on. Enabling automation is a non-goal.
* User experience is important. Features should be intuitive and discoverable, but efficiency is more important than discoverability. More advanced features should be hidden, and primary features should be central. Performance is key.
* Automatically provide as much contextual information as possible, without overwhelming the user. Information that is presented automatically when needed is a lot more useful than information you need to explicitly request. For example, provide autocomplete for every input.
* Provide easy access to the raw Salesforce API. Enhance the interaction in a way that does not break the core use case, if our enhancements fails. For example, ensure we can display the result of a data export even if we cannot parse the SOQL query.
* It is fine to implement features that are already available in the core Salesforce UI, if we can make it easier, smarter or faster.
* Ensure that it works for as many users as possible. (for system administrators, for standard users, with person accounts, with multi currency, with large data volumes, with professional edition, on a slow network etc.)
* Be conservative about the number and complexity of Salesforce API requests we make, but don't sacrifice the other principles to do so.
* Focus on system administrators, developers and integrators.

About
-----
By Søren Krabbe and Jesper Kristensen

License
-----
MIT


CHANGES.md
Version 1.15
===========

General
=======
* Update to Salesforce API v 59.0 (Winter'24)
* Update packages

Version 1.14
===========

General
=======
* Update to Salesforce API v 56.0 (Winter'23)

Version 1.13
===========

General
=======
* Update to Salesforce API v 55.0 (Summer'22)

Version 1.12
===========

General
-------
* Update to Salesforce API v 51.0 (Spring '21)

Version 1.11
===========

General
-------
* Make inspector available on Visualforce pages on new visualforce.com domain. See #143

Org Limits
----------
* Displays "consumed" count

Version 1.10
===========

General
-------
* Update to Salesforce API v 48.0

Version 1.9
===========

Inspector menu
--------------
* Fix a bug fix hiding the "show field metadata" button (#150)

Version 1.8
===========

Inspector menu
--------------
* Added user search aspect to simplify access to detailed user data and "login as".

Version 1.7
===========

General
-------
* Update to Salesforce API v 47.0

Inspector menu
--------------
* A new link to switch in and out of Salesforce Setup, where you can choose to open in a new tab or not.

Show all data
-------------
* Fixed a bug causing errors when viewing some special objects.
* Link to Salesforce Setup in both Classic and Lightning Experience.
* Use default values for blank fields when creating a new record. This avoids the error message that OwnerId is required but missing.

Data import
-----------
* Save import options in your excel sheet, so you can update the same data again and again with a single copy-paste.


Version 1.6
===========

General
-------
* Update to Salesforce API v 45.0
* Support for cloudforce.com orgs

Show all data
-------------
* Buttons to Create, delete and clone records

Data export
-----------
* Keyboard shortcut to do export (ctrl+enter)
* Fixes saved query selection

Data import
-----------
* Wider import fields

Version 1.5
===========

General
-------
* Update to Salesforce API v 43.0

Inspector menu
--------------
* Show record details - currently for objects with record types only
* Link to LEX object manager/setup for object in focus

Version 1.4
===========

Inspector menu
--------------
* Support for Spring '18 LEX URL format (https://docs.releasenotes.salesforce.com/en-us/spring18/release-notes/rn_general_enhanced_urls_cruc.htm)

Version 1.3
===========

General
-------
* Rewritten the implementation of Data Export and Data Import, in order to comply with the updated version of Mozilla's add-ons policy.

Version 1.2
===========

General
-------
* Update API versoin to Spring 17.

Inspector menu
--------------
* Use the autocomplete to find object API names, labels and ID prefixes.
* View some information about the selected record or object directly in the menu.
* Inspect objects in the Tooling API and objects you don't have read access to.
* When viewing a Deployment Status, a new button allows you to get all the details of the deployment.
* The Explore API button is now visible everywhere.

Show all data
-------------
* The Type column has more information. (required, unique, auto number etc.)
* Add your own columns, (for example a column showing the formula of formula fields, or a collumn that tells which fields can be used as a filter.) for both fields and relationships.
* The "Advanced filter" option is more discoverable now.
* New button to start data export for the shown object.
* New button to edit the page layout for the shown record.
* Better handling of objects that share a common ID prefix or is available with both the regular API and the Tooling API.

Data export
-----------
* Save your favourite SOQL queries.
* The query history remembers if queries were done with the Tooling API or not.
* Fixed right clicking on IDs in the exported data.

Data import
-----------
* Fix for importing data from Excel on Mac into Chrome.

Org Limits
----------
* View how much of your org's limits you are currently using.

Download Metadata
-----------------
* Download all your org's Apex classes, Visualforce pages, objects, fields, validation rules, workflow rules, reports and much more. Use it for backup, or if you want to search for any place a particular item is used, or for many other purposes.

API Explorer
------------
* Choose between showing the result for easy viewing or for easy copying.
* Make SOAP requests.
* Make REST requests for any HTTP method.
* Edit any API request before sending.

Version 1.1
============

General
-------
* Update API versoin to Winter 17.
* Find the current page's record ID for Visualforce pages that store the record ID in a non-standard parameter name.

Data import
-----------
* Don't make describe calls in an infinite loop when Salesforce returns an error (Salesforce Winter 17 Tooling API has a number objects starting with autogen__ that don't work properly).

Version 1.0
============

General
-------
* The Inspector is now shown in regular tabs instead of popups. You can now choose if you want to open a link in the same tab (the default), or a new tab/window, using normal browser menus and shortcuts. Previously every link opened a new popup window.
* Restyled the Inspector menu to use Lightning Design. Restyling the rest will come later.
* Switched to a more robust API for getting the Salesforce session ID. It now works with all session security settings, and it works in Lightning Experience.
* Added a logo/icon.
* The salesforce hostname is now visible as a parameter in the URL bar.
* If you have an outdated browser version that is not supported by the latest version of Salesforce Inspector, Salesforce Inspector will not autoupdate.
* Updated API version to Summer 16.

Show all data
-------------
* When copy-pasting a value, there is no longer extra white-space at the beginning and end of the copied text.


addon/limits.js
/* global React ReactDOM */
import {sfConn, apiVersion} from "./inspector.js";
/* global initButton */

class Model {
  constructor(sfHost) {
    this.reactCallback = null;
    this.sfLink = "https://" + sfHost;
    this.spinnerCount = 0;
    this.allLimitData = [];
    this.errorMessages = [];
    this.userInfo = '...';
    this.spinFor(sfConn.soap(sfConn.wsdl(apiVersion, "Partner"), "getUserInfo", {}).then(res => {
      this.userInfo = res.userFullName + " / " + res.userName + " / " + res.organizationName;
    }));
  }

  didUpdate(cb) {
    if (this.reactCallback) {
      this.reactCallback(cb);
    }
  }

  spinFor(promise) {
    this.spinnerCount++;
    promise
      .catch(err => {
        console.error("spinFor", err);
      })
      .then(() => {
        this.spinnerCount--;
        this.didUpdate();
      })
      .catch(err => console.log("error handling failed", err));
  }
  
  spinFor2(actionName, promise, cb) {
    this.spinnerCount++;
    promise
      .then(res => {
        this.spinnerCount--;
        cb(res);
        this.didUpdate();
      })
      .catch(err => {
        console.error(err);
        this.errorMessages.push("Error " + actionName + ": " + err.message);
        this.spinnerCount--;
        this.didUpdate();
      })
      .catch(err => console.log("error handling failed", err));
  }

  title() {
    return "Org Limits";
  }

  setLimitsData(res) {
    let self = this;
    this.allLimitData = [];

    Object.keys(res).forEach((key) => {
      self.allLimitData.push({
        key,
        "label": self.humanizeName(key),
        "description": "...",
        "max": res[key].Max,
        "remaining": res[key].Remaining
      });
    });
  }

  humanizeName(name) {
    return name.replace(/([A-Z])/g, " $1"); //TODO: Improve
  }

  startLoading() {
    let limitsPromise = sfConn.rest("/services/data/v" + apiVersion + "/" + "limits");

    this.spinFor2("describing global", limitsPromise, (res) => {
      this.setLimitsData(res);
    });
    
  }
}


let h = React.createElement;

class LimitData extends React.Component {
  render() {
    return (
      h("figure", {},
        h("div", {
          className: "gauge"
        },
        h("div", {
          className: "meter",
          ref: "meter"
        },
        ""
        ),
        h("div", {
          className: "meter-value-container"
        },
        h("div", {
          className: "meter-value"
        }, Math.round((1 - (this.props.remaining / this.props.max)) * 100) + "%")
        )
        ),
        h("figcaption", {}, this.props.label,
          h("div", {}, (this.props.max - this.props.remaining).toLocaleString() + " of " + (this.props.max).toLocaleString() + " consumed",
            h("br", {}), "(" + (this.props.remaining).toLocaleString() + " left)"
          ),
        )
      )
    );
  }
  componentDidMount() {
    // Animate gauge to relevant value
    let targetDegree = (this.props.max == 0) ? "180deg" : ((1 - (this.props.remaining / this.props.max)) * 180) + "deg"; //180deg = 100%, 0deg = 0%
    this.refs.meter.animate([{
      transform: "rotate(0deg)"
    }, {
      transform: "rotate(" + targetDegree + ")"
    }], {
      duration: 1000,
      fill: "both"
    });
  }
}

class App extends React.Component {
  render() {
    let vm = this.props.vm;
    document.title = vm.title();
    return (
      h("div", {},
        h("div", {id: "user-info"},
          h("a", {href: vm.sfLink, className: "sf-link"},
            h("svg", {viewBox: "0 0 24 24"},
              h("path", {d: "M18.9 12.3h-1.5v6.6c0 .2-.1.3-.3.3h-3c-.2 0-.3-.1-.3-.3v-5.1h-3.6v5.1c0 .2-.1.3-.3.3h-3c-.2 0-.3-.1-.3-.3v-6.6H5.1c-.1 0-.3-.1-.3-.2s0-.2.1-.3l6.9-7c.1-.1.3-.1.4 0l7 7v.3c0 .1-.2.2-.3.2z"})
            ),
            " Salesforce Home"
          ),
          h("h1", {}, "Org Limits"),
          h("span", {}, " / " + vm.userInfo),
          h("div", {className: "flex-right"},
            h("div", {id: "spinner", role: "status", className: "slds-spinner slds-spinner_small slds-spinner_inline", hidden: vm.spinnerCount == 0},
              h("span", {className: "slds-assistive-text"}),
              h("div", {className: "slds-spinner__dot-a"}),
              h("div", {className: "slds-spinner__dot-b"}),
            ),
            h("a", {href: "#", id: "help-btn", title: "Export Help", onClick: this.onToggleHelp}, 
              h("div", {className: "icon"})
            ),
          ),
        ),
        h("div", {
          className: "body"
        },
        vm.allLimitData.map(limitData =>
          h(LimitData, limitData)
        )
        )
      )
    );
  }
}


{

  let args = new URLSearchParams(location.search.slice(1));
  let sfHost = args.get("host");
  initButton(sfHost, true);
  sfConn.getSession(sfHost).then(() => {

    let root = document.getElementById("root");
    let vm = new Model(sfHost);
    vm.startLoading(args);
    vm.reactCallback = cb => {
      ReactDOM.render(h(App, {
        vm
      }), root, cb);
    };
    ReactDOM.render(h(App, {
      vm
    }), root);

  });

  {
    let isDragging = false;
    document.body.onmousedown = () => {
      isDragging = false;
    };
    document.body.onmousemove = e => {
      if (e.movementX || e.movementY) {
        isDragging = true;
      }
    };
    document.body.onclick = e => {
      if (!e.target.closest("a") && !isDragging) {
        let el = e.target.closest(".quick-select");
        if (el) {
          getSelection().selectAllChildren(el);
        }
      }
    };
  }

}
